### PR TITLE
chore: partially remove attrs II

### PR DESCRIPTION
**Pull request by gnufede**
## Description

This PR some of `attrs` usage, substituting it with `dataclasses` or plain Python classes if slots was used.

## Checklist

- [x] The PR description includes an overview of the change
- [x] The PR description articulates the motivation for the change
- [x] The change includes tests OR the PR description describes a testing strategy
- [x] The PR description notes risks associated with the change, if any
- [x] Newly-added code is easy to change
- [x] The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- [x] The change includes or references documentation updates if necessary
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [ ] Title is accurate
- [ ] All changes are related to the pull request's stated goal
- [ ] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [ ] Testing strategy adequately addresses listed risks
- [ ] Newly-added code is easy to change
- [ ] Release note makes sense to a user of the library
- [ ] If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [ ] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: ddtrace/internal/rate_limiter.py [hunk 1]
```diff
@@ -64,12 +63,12 @@ class BudgetRateLimiterWithJitter:
     limit_rate: float
     tau: float = 1.0
     raise_on_exceed: bool = True
-    on_exceed: Optional[Callable] = None
+    on_exceed: Optional[Callable[[], None]] = None
```
Comment (gnufede) [inline ddtrace/internal/rate_limiter.py:66]:
```suggestion
    on_exceed: Optional[Callable] = None
```

File: ddtrace/bootstrap/sitecustomize.py [hunk 2]
```diff
@@ -76,6 +76,7 @@ def drop(module_name):
             "sre_constants",  # imported by re at runtime
             "logging",
             "attr",
+            "dataclasses",
```
Comment (brettlangdon) [inline ddtrace/bootstrap/sitecustomize.py:79]: hmm, why do we need this here?

File: ddtrace/contrib/logging/patch.py [hunk 3]
```diff
@@ -32,13 +30,20 @@ def get_version():
...
-    env = attr.ib(type=str)
+class DDLogRecord:
+    trace_id: int
+    span_id: int
+    service: str
+    version: str
+    env: str
+    __slots__ = ("trace_id", "span_id", "service", "version", "env")
+
+    def __init__(self, trace_id, span_id, service, version, env):
```
Comment (brettlangdon) [inline ddtrace/contrib/logging/patch.py:41]:
```suggestion
    def __init__(self, trace_id: int, span_id: int, service: str, version: str, env: str):
```

File: ddtrace/settings/config.py [hunk 4]
```diff
@@ -545,7 +545,7 @@ def __init__(self):
             # https://github.com/open-telemetry/opentelemetry-python/blob/v1.16.0/opentelemetry-api/src/opentelemetry/context/__init__.py#L53
             os.environ["OTEL_PYTHON_CONTEXT"] = "ddcontextvars_context"
         self._ddtrace_bootstrapped = False
-        self._subscriptions = []  # type: List[Tuple[List[str], Callable[[Config, List[str]], None]]]
+        self._subscriptions: List[Tuple[List[str], Callable[[Config, List[str]], None]]] = []
```
Comment (brettlangdon) [inline ddtrace/settings/config.py:548]: unrelated change?

**Comment by gnufede** [comment]
Split into separate PRs
